### ISSUE TITLE
Unable to use any ElasticSearch API after the service is up

### ISSUE THREAD

**Issue by fvcaputo** [localstack/localstack#3081]
# Type of request: This is a ...
[x] bug report
[ ] feature request

# Detailed description
I'm running localstack through docker, image `localstack/localstack-full:0.11.6` with mostly default configuration (using services sqs and es):
![image](https://user-images.githubusercontent.com/13631451/95380175-bf392800-08bc-11eb-96aa-ade5af198b2f.png)

Docker compose runs everything correctly, and localstack is ready. After it's already up I run the command:

```sh
aws --endpoint-url http://localhost:4566 es create-elasticsearch-domain --domain-name domain-test
```

![image](https://user-images.githubusercontent.com/13631451/95379829-46d26700-08bc-11eb-99d1-691416a7fb36.png)

To create a domain. This works correctly too, and after a few minutes I see the log that service is running at `0.0.0.0:4571`:
![image](https://user-images.githubusercontent.com/13631451/95379844-4c2fb180-08bc-11eb-84fe-7fbe26d5ba93.png)

But after that I cannot seem to correctly utilize elasticsearch at all. Any request made to it returns a 500 error code, for instance the request that would create an index:

```sh
$ curl -X PUT "http://localhost:4571/index" -v                                              
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to localhost (127.0.0.1) port 4571 (#0)
> PUT /index HTTP/1.1
> Host: localhost:4571
> User-Agent: curl/7.58.0
> Accept: */*
> 
< HTTP/1.1 500 
< content-type: text/html; charset=utf-8
< content-length: 2
< date: Wed, 07 Oct 2020 19:53:09 GMT
< server: hypercorn-h11
< 
* Connection #0 to host localhost left intact
{}
```

And no logs seem to appear on localstack running on docker.

## Expected behavior
After the service is up I should be able to correctly use ElasticSearch's apis.

## Actual behavior
Only 500 status code from every request.

# Steps to reproduce
## Command used to start LocalStack
docker-comose up

## Client code (AWS SDK code snippet, or sequence of "awslocal" commands)
AWS commands that use ElasticSeach Service do work:

```sh
$ aws es describe-elasticsearch-domain --domain-name domain-test --endpoint-url http://localhost:4566
```

But requests to the elasticsearch search engine do not work.

┆Issue is synchronized with this [Jira Bug](https://localstack.atlassian.net/browse/LOC-280) by [Unito](https://www.unito.io/learn-more)

**Comment by fvcaputo** [comment]
Oh, I have also tried with different localstack versions, couldn't really get any to work. But version `localstack/localstack-full:0.11.5` has an interesting behaviour, the same curl command returns a 200:
```sh
$ curl -X PUT "http://localhost:4571/index" -v      
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to localhost (127.0.0.1) port 4571 (#0)
> PUT /index HTTP/1.1
> Host: localhost:4571
> User-Agent: curl/7.58.0
> Accept: */*
> 
< HTTP/1.1 200 
< content-type: text/html; charset=utf-8
< content-length: 2
< date: Wed, 07 Oct 2020 20:04:35 GMT
< server: hypercorn-h11
< 
* Connection #0 to host localhost left intact
{}%
```
But the command doesn't really work, that is not the expected return (and the same thing happens with any curl request, a 200 with a return of an empty object `{}`).

And in this version, then the request is made, some error logs do appear from localstack:
```sh
localstack_1                     | 2020-10-07T20:05:31:WARNING:bootstrap.py: Thread run method <function AdaptiveThreadPool.submit.<locals>._run at 0x7fbc0dc17e50>(None) failed: MyHTTPConnectionPool(host='127.0.0.1', port=41957): Max retries exceeded with url: /index (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fbc0f9228b0>: Failed to establish a new connection: [Errno 111] Connection refused')) Traceback (most recent call last):
localstack_1                     |   File "/opt/code/localstack/.venv/lib/python3.8/site-packages/urllib3/connection.py", line 159, in _new_conn
localstack_1                     |     conn = connection.create_connection(
localstack_1                     |   File "/opt/code/localstack/.venv/lib/python3.8/site-packages/urllib3/util/connection.py", line 84, in create_connection
localstack_1                     |     raise err
localstack_1                     |   File "/opt/code/localstack/.venv/lib/python3.8/site-packages/urllib3/util/connection.py", line 74, in create_connection
localstack_1                     |     sock.connect(sa)
localstack_1                     | ConnectionRefusedError: [Errno 111] Connection refused
localstack_1                     | 
localstack_1                     | During handling of the above exception, another exception occurred:
localstack_1                     | 
localstack_1                     | Traceback (most recent call last):
localstack_1                     |   File "/opt/code/localstack/.venv/lib/python3.8/site-packages/urllib3/connectionpool.py", line 670, in urlopen
localstack_1                     |     httplib_response = self._make_request(
localstack_1                     |   File "/opt/code/localstack/.venv/lib/python3.8/site-packages/urllib3/connectionpool.py", line 392, in _make_request
localstack_1                     |     conn.request(method, url, **httplib_request_kw)
localstack_1                     |   File "/usr/lib/python3.8/http/client.py", line 1230, in request
localstack_1                     |     self._send_request(method, url, body, headers, encode_chunked)
localstack_1                     |   File "/usr/lib/python3.8/http/client.py", line 1276, in _send_request
localstack_1                     |     self.endheaders(body, encode_chunked=encode_chunked)
localstack_1                     |   File "/usr/lib/python3.8/http/client.py", line 1225, in endheaders
localstack_1                     |     self._send_output(message_body, encode_chunked=encode_chunked)
localstack_1                     |   File "/usr/lib/python3.8/http/client.py", line 1004, in _send_output
localstack_1                     |     self.send(msg)
localstack_1                     |   File "/usr/lib/python3.8/http/client.py", line 944, in send
localstack_1                     |     self.connect()
localstack_1                     |   File "/opt/code/localstack/.venv/lib/python3.8/site-packages/urllib3/connection.py", line 187, in connect
localstack_1                     |     conn = self._new_conn()
localstack_1                     |   File "/opt/code/localstack/.venv/lib/python3.8/site-packages/urllib3/connection.py", line 171, in _new_conn
localstack_1                     |     raise NewConnectionError(
localstack_1                     | urllib3.exceptions.NewConnectionError: <urllib3.connection.HTTPConnection object at 0x7fbc0f9228b0>: Failed to establish a new connection: [Errno 111] Connection refused
localstack_1                     | 
localstack_1                     | During handling of the above exception, another exception occurred:
localstack_1                     | 
localstack_1                     | Traceback (most recent call last):
localstack_1                     |   File "/opt/code/localstack/.venv/lib/python3.8/site-packages/requests/adapters.py", line 439, in send
localstack_1                     |     resp = conn.urlopen(
localstack_1                     |   File "/opt/code/localstack/.venv/lib/python3.8/site-packages/urllib3/connectionpool.py", line 726, in urlopen
localstack_1                     |     retries = retries.increment(
localstack_1                     |   File "/opt/code/localstack/.venv/lib/python3.8/site-packages/urllib3/util/retry.py", line 439, in increment
localstack_1                     |     raise MaxRetryError(_pool, url, error or ResponseError(cause))
localstack_1                     | urllib3.exceptions.MaxRetryError: MyHTTPConnectionPool(host='127.0.0.1', port=41957): Max retries exceeded with url: /index (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fbc0f9228b0>: Failed to establish a new connection: [Errno 111] Connection refused'))
localstack_1                     | 
localstack_1                     | During handling of the above exception, another exception occurred:
localstack_1                     | 
localstack_1                     | Traceback (most recent call last):
localstack_1                     |   File "/opt/code/localstack/localstack/utils/bootstrap.py", line 521, in run
localstack_1                     |     result = self.func(self.params)
localstack_1                     |   File "/opt/code/localstack/localstack/utils/async_utils.py", line 28, in _run
localstack_1                     |     return fn(*args, **kwargs)
localstack_1                     |   File "/opt/code/localstack/localstack/services/generic_proxy.py", line 555, in handler
localstack_1                     |     response = modify_and_forward(method=method, path=path_with_params, data_bytes=data, headers=headers,
localstack_1                     |   File "/opt/code/localstack/localstack/services/generic_proxy.py", line 372, in modify_and_forward
localstack_1                     |     response = requests_method(request_url, data=data_to_send,
localstack_1                     |   File "/opt/code/localstack/.venv/lib/python3.8/site-packages/requests/api.py", line 134, in put
localstack_1                     |     return request('put', url, data=data, **kwargs)
localstack_1                     |   File "/opt/code/localstack/.venv/lib/python3.8/site-packages/requests/api.py", line 61, in request
localstack_1                     |     return session.request(method=method, url=url, **kwargs)
localstack_1                     |   File "/opt/code/localstack/.venv/lib/python3.8/site-packages/requests/sessions.py", line 530, in request
localstack_1                     |     resp = self.send(prep, **send_kwargs)
localstack_1                     |   File "/opt/code/localstack/.venv/lib/python3.8/site-packages/requests/sessions.py", line 643, in send
localstack_1                     |     r = adapter.send(request, **kwargs)
localstack_1                     |   File "/opt/code/localstack/.venv/lib/python3.8/site-packages/requests/adapters.py", line 516, in send
localstack_1                     |     raise ConnectionError(e, request=request)
localstack_1                     | requests.exceptions.ConnectionError: MyHTTPConnectionPool(host='127.0.0.1', port=41957): Max retries exceeded with url: /index (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fbc0f9228b0>: Failed to establish a new connection: [Errno 111] Connection refused'))
localstack_1                     | 
```

**Comment by whummer** [comment]
Thanks for reporting @fvcaputo . Can you please verify if this issue persists in the `latest` version of the Docker image?

The following seems to work:
```
$ awslocal es create-elasticsearch-domain --domain-name domain-test
...
        "Endpoint": "http://localhost:4571",
...
$ awslocal es describe-elasticsearch-domain --domain-name domain-test | grep Created
        "Created": true,
$ curl -X PUT "http://localhost:4571/index1"
{"acknowledged":true,"shards_acknowledged":true,"index":"index1"}
```
Please note that it is required to poll the ES domain status (via `describe-elasticsearch-domain`) until `Created` is true.

Hope that helps - please keep us posted.

**Comment by mruz** [comment]
On the 0.11.6 I keep seeing:
>         "Created": false,

**Comment by mattsoftware** [comment]
I am seeing the same thing. This was working for me until I pulled the latest localstack-full image. Lots of exceptions in the container logs as per the above comments too.

**Comment by mattsoftware** [comment]
Roling back to the old container seems to not have solved the issue. After trying several tags I went back to :latest and now my domains are created:true. Perhaps it needed a couple of boots to sort out the existing data?

Anyway, its working for me again on the :latest tag. Thanks.

**Comment by sivnag** [comment]
@fvcaputo, I am facing exact same issue. I have trying various versions of localstack and localstack-full and results are identical.
I have attached logs from latest version of localstack-full that was available at the time of this writing (DIGEST:sha256:700d815065d6e47ef114f525d2f7e0221749087c7fd1e55755a4a924d516e996)

[localstack.error.log](https://github.com/localstack/localstack/files/5544452/localstack.error.log)

@mattsoftware, can you please share the digest information for the localstack image that you have found to be working?

docker images --digests | grep localstack

**Comment by sivnag** [comment]
> Thanks for reporting @fvcaputo . Can you please verify if this issue persists in the `latest` version of the Docker image?
> 
> The following seems to work:
> 
> ```
> $ awslocal es create-elasticsearch-domain --domain-name domain-test
> ...
>         "Endpoint": "http://localhost:4571",
> ...
> $ awslocal es describe-elasticsearch-domain --domain-name domain-test | grep Created
>         "Created": true,
> $ curl -X PUT "http://localhost:4571/index1"
> {"acknowledged":true,"shards_acknowledged":true,"index":"index1"}
> ```
> 
> Please note that it is required to poll the ES domain status (via `describe-elasticsearch-domain`) until `Created` is true.
> 
> Hope that helps - please keep us posted.

@whummer can you please let me know approximately how long should we poll for? I seem to get ' "Created": false ' irrespective of duration of polling. Also there are lots of 'connection refused' errors. I have added my comment in this thread with all information about my env + error logs.

**Comment by sivnag** [comment]
exporting DEBUG=1, brought some more logs which showed that there was a permission error trying to access /tmp/localstack/es_backup folder. Once I created es_backup folder in host linux machine, same folder was available in the mounted volume inside the docker container and with this change, create-elasticsearch-domain + describe-elasticsearch-domain eventually returned "Created": true after about 10 seconds.

However, loads of "Connection refused errors" are still present in the logs (similar to what I had uploaded earlier)
localstack_main | requests.exceptions.ConnectionError: MyHTTPConnectionPool(host='127.0.0.1', port=36097): Max retries exceeded with url: /_cat/aliases (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f7e3558b610>: Failed to establish a new connection: [Errno 111] Connection refused'))

**Comment by garethmdavies** [comment]
Just hit this same issue when moving to 0.12.2 from 0.11.2 to take advantage of some S3 CORS changes. Updated config to use new edge port but unable to successfully create the ES domain. See the same connection refused issues as above.

**Comment by FAYiEKcbD0XFqF2QK2E4viAHg8rMm2VbjYKdjTg** [comment]
I am seeing the same issues as well. Was working for me before- start docker, create domain, curl put some data. Now it's empty responses against port 4571.

**Comment by fuasahina** [comment]
I am seeing the same issues as well. too . I use lastest version 12.2 images on docker. 　installed `local Elastic Search` versions is `6.7.0` and ES endpoint url  `4571`

**Comment by whummer** [comment]
Thanks for the update @fvcaputo @fuasahina @FAYiEKcbD0XFqF2QK2E4viAHg8rMm2VbjYKdjTg @garethmdavies @sivnag @mattsoftware @mruz . A couple of fixes have been pushed in #3565 and we believe (hope) that this issue should now be fixed. Can you please pull the `latest` Docker image and give it another try? Please report here if the problem persists. Thanks!
